src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -1015,7 +1015,11 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
          (this hasFlag FINAL | PACKAGE)
       || isModuleOrModuleClass && (isTopLevel || !settings.overrideObjects)
       || isTerm && (isPrivate || isLocalToBlock || (hasAllFlags(notPRIVATE | METHOD) && !hasFlag(DEFERRED)))
-      || isClass && originalOwner.isTerm && children.isEmpty // we track known subclasses of term-owned classes, use that infer finality
+      || isClass && !isRefinementClass && originalOwner.isTerm && children.isEmpty // we track known subclasses of term-owned classes, use that infer finality
+      // don't look at owner for refinement classes (it's basically arbitrary) -- instead,
+      // it suffices for one parent of an intersection to be final, for the resulting type to be final
+      // any subclass of the refinement would have to be a subclass of that final parent, which is not allowed
+      || isRefinementClass && info.parents.exists { _.typeSymbol.isEffectivelyFinal }
```

Comment:
This seems a little expensive as this method is called from some hot places, e.g. `rebind`. 

I noticed a [small slowdown](https://scala-ci.typesafe.com/grafana/dashboard/db/scala-benchmark?orgId=1&from=1535480978910&to=1536045180162&var-branch=2.12.x&var-source=All&var-bench=HotScalacBenchmark.compile&var-host=scalabench%40scalabench%40) in benchmarks.

Could we move this part out to the use site in `RefChecks` rather than baking it to `isEffectivelyFinal`?

Reply:
#7177